### PR TITLE
[ENH] Add the ability to find Proper Possibly Directed Paths

### LINKED ISSUE #111: [ENH] Add the ability to find Proper Possibly Directed Paths in a Graph
**Is your feature request related to a problem? Please describe.**
This is in continuation of PR [#1148](https://github.com/py-why/dowhy/pull/1148) in Dowhy. First PR in an effort to provide seamless integration with Dowhy.

**Describe the solution you'd like**
A couple of functions that can take a graphs and find Proper Possibly Directed Paths in a graph.

### PR DESCRIPTION
**Pull request by aryan26roy**
Add the ability to find Proper Possibly Directed Paths to allow integration with DoWhy.

Fixes #111 

Changes proposed in this pull request:

- Add the necessary functions to find Proper Possibly Directed Paths in a Graph.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
The CIs should now work. You can also test locally ofc with pytest.

**Comment by aryan26roy** [comment]
@adam2392 I have completed the implementation and it seems to be correct. However, there is one problem. The output is a list of dictionaries. And the dictionaries can be in any order inside the list which is why the test may fail right now. How do I compare them in an order agnostic way?

**Comment by adam2392** [comment]
Is there any inspiration from network's method for testing `all_simple_paths`?

https://github.com/networkx/networkx/blob/main/networkx/algorithms/tests/test_simple_paths.py

**Comment by aryan26roy** [comment]
They are returning a set of tuples. I am assuming the order of tuples inside a set won't matter during comparison because of it's hashibality. However, I cannot store dictionaries inside sets. Either I change my implementation to store paths in sets(tuple()) or I can do the tests using the brute force method I described above.

What do you want me to do?

**Comment by adam2392** [comment]
Is there any technical reason to do dictionaries instead of triples that I'm missing?

File: pywhy_graphs/algorithms/generic.py [hunk 1]
```diff
@@ -855,3 +856,99 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
                 else:
                     vstructs.add((p1, node, p2))  # type: ignore
     return vstructs
+
+
+def check_back_arrow(G: ADMG, X, Y: set):
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:861]: Let's make this a private function?

Also, document that this checks if any node in `Y` points back to `X`, where `X` is a node, not a set? Or did I misunderstand that?

File: pywhy_graphs/algorithms/generic.py [hunk 2]
```diff
@@ -855,3 +856,99 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+            temp_paths = paths[i].copy()
+            temp_paths[len(temp_paths)] = elem
+            temp_arr.append(temp_paths)
+
+        new_paths.extend(recursively_find_pd_paths(G, X, temp_arr, Y))
+
+    return new_paths
+
+
+def possibly_directed_path(G, X: Optional[Set] = None, Y: Optional[Set] = None):
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:937]: Can we add a docstring?

File: pywhy_graphs/algorithms/generic.py [hunk 3]
```diff
@@ -855,3 +856,99 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+    else:
+        nbr_temp = G.neighbors(X)
+        nbr_possible = check_back_arrow(nbr_temp)
+        x_neighbors = []
+
+        for elem in nbr_possible:
+            temp = dict()
+            temp[0] = X
+            temp[1] = elem
+            x_neighbors.append(temp)
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:950]: I'm confused why there's two paths depending on the type of X. Can't we just use `{G.neighbors(x) for x in X}` to get a set of all neighbors of X if X is a set?

**Review by adam2392 — COMMENTED** [review]
Just left a few comments to hopefully push in the right direction

**Comment by aryan26roy** [comment]
Sorry for the late reply.

> Is there any technical reason to do dictionaries instead of triples that I'm missing?

Not really, I just did not think of using tuples at the time. I think it would be easy to go from List<Dict> to Set<Tuples>. A better approach in terms of memory as well.

File: pywhy_graphs/algorithms/generic.py (hunk 1 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:861]: Was about to do it. I just wanted to know your opinion on how to test it, the PR is not ready for a full review yet. I should have been more explicit. 😅

File: pywhy_graphs/algorithms/generic.py (hunk 2 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:937]: Yep!

File: pywhy_graphs/algorithms/generic.py (hunk 3 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:950]: I did not know that G.neighbors(x) can take a set as an argument as well. Thanks for pointing this out.

Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:949]: I tried running G.neighbors on a set of nodes and it errors out:

`TypeError: unhashable type: 'set'`

Is it supposed to throw this error or is it a bug?

File: pywhy_graphs/algorithms/generic.py (hunk 1 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:861]: > Also, document that this checks if any node in `Y` points back to `X`, where `X` is a node, not a set? Or did I misunderstand that?

Yes you understood that correctly, it is never called for a set X.

File: pywhy_graphs/algorithms/generic.py (hunk 3 above)
Comment (adam2392) [reply pywhy_graphs/algorithms/generic.py:949]: Ah yeah that is correct. I meant can we just have a one liner that iterates over the nodes in X and adds to a nbr set? The resulting nbr set is just all the nodes that are nbrs of X.

Or is there anything more nuanced happening?

**Comment by aryan26roy** [comment]
@adam2392 I just realised, since in this implementation, all the first nodes of all the paths are in X, this finds all the proper possibly directed paths, doesn't it?

File: pywhy_graphs/algorithms/generic.py (hunk 3 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py]: I wrote that function to account for the fact that the neighbor of a member of X might be another member of X. Given the following definition of a proper path:
>  A path from X to Y is proper (wrt X) if only its first node is in X

Am I wrong? Or do we need to to do this check?

**Comment by adam2392** [comment]
Perhaps it's easier to implement just the function wrt a single node rather than a set of nodes. Do we need the functionality with X as a set? Wdyt?

**Comment by aryan26roy** [comment]
The paper assumes everywhere that X will be a set. And the current implementation already covers and tests X being a set. It seems to me that changing the implementation now to only support a single node in X would be wasted effort.

**Comment by adam2392** [comment]
> @adam2392 I just realised, since in this implementation, all the first nodes of all the paths are in X, this finds all the proper possibly directed paths, doesn't it?

Yes I believe so.

**Comment by aryan26roy** [comment]
@adam2392 I believe the PR is complete. Can you do one last round of review? (PS. The failing test is due to some environment issue with some existing test)

File: pywhy_graphs/algorithms/generic.py [hunk 4]
```diff
@@ -855,3 +856,161 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+    """Find all the possibly directed paths in a graph.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    X : Set
+        Source.
+    Y : Set
+        Destination
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:992]: What if both are None? I don't think None should even be allowed, but lmk if I misunderstood

File: pywhy_graphs/algorithms/generic.py [hunk 5]
```diff
@@ -855,3 +856,161 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+    for elem in Y:
+        if not (
+            G.has_edge(X, elem, G.bidirected_edge_name) or G.has_edge(elem, X, G.directed_edge_name)
+        ):
+            out.update(elem)
+
+    return out
+
+
+def _get_X_neighbors(G, X: set):
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:890]:
```suggestion
def _get_neighbors_of_set(G, X: set):
```

to be more explicit?

File: pywhy_graphs/algorithms/generic.py [hunk 6]
```diff
@@ -855,3 +856,161 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+    """Retrieve all the neigbors of X when X has more than one element.
+
```
Comment (adam2392) [inline pywhy_graphs/algorithms/generic.py:892]: You can add a note here saying if `X` is not a set, then one can simply use `graph.neighbors(X)`

**Review by adam2392 — COMMENTED** [review]
Few short comments, but otw, it's looking good

File: pywhy_graphs/algorithms/tests/test_generic.py [hunk 7]
```diff
@@ -496,3 +496,179 @@ def test_all_vstructures():
     # Assert that the returned values are as expected
     assert len(v_structs_edges) == 0
     assert len(v_structs_tuples) == 0
+
+
+def test_proper_possibly_directed():
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/tests/test_generic.py:501]: Separate them in separate functions or through comments?

File: pywhy_graphs/algorithms/generic.py [hunk 8]
```diff
@@ -855,3 +856,161 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+            temp_paths = elem
+            temp_paths = temp_paths + (nbr,)
+            temp_set.add(temp_paths)
+
+        new_paths.update(_recursively_find_pd_paths(G, X, temp_set, Y))
+
+    return new_paths
+
+
+def proper_possibly_directed_path(G, X: Optional[Set] = None, Y: Optional[Set] = None):
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:982]: Got it!

File: pywhy_graphs/algorithms/generic.py (hunk 4 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:992]: Yes you are right, neither can ever be None. Maybe I should return an empty set whenever either one of them is None.

File: pywhy_graphs/algorithms/generic.py (hunk 5 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:890]: Sure.

File: pywhy_graphs/algorithms/generic.py (hunk 6 above)
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py:892]: Alright.

File: pywhy_graphs/algorithms/tests/test_generic.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adam2392) [reply pywhy_graphs/algorithms/tests/test_generic.py:501]: Functions

File: pywhy_graphs/algorithms/generic.py (hunk 4 above)
Comment (adam2392) [reply pywhy_graphs/algorithms/generic.py:992]: Why not just not have default argument?

**Comment by aryan26roy** [comment]
@adam2392 I have incorporated all of your suggestions. Can you take another look?

**Review by adam2392 — APPROVED** [review]
LGTM except two minor nits

File: pywhy_graphs/algorithms/generic.py [hunk 9]
```diff
@@ -855,3 +856,187 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+    graph.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    X : Set
+        Source.
+    Y : Set
+        Destination
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py]: Oh yeah!

File: pywhy_graphs/algorithms/generic.py [hunk 10]
```diff
@@ -855,3 +856,187 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
...
+        x_neighbors = []
+
+        for elem in nbr_possible:
+            temp = dict()
+            temp[0] = X
+            temp[1] = elem
+            x_neighbors.append(temp)
+
+    path_list = _recursively_find_pd_paths(G, X, x_neighbors, Y)
+    print(path_list)
```
[the remark this answers is not in the document]
Comment (aryan26roy) [reply pywhy_graphs/algorithms/generic.py]: Removed.

**Comment by aryan26roy** [comment]
Done!

**Comment by aryan26roy** [comment]
@adam2392 This can be optimised further by switching to a stack based implementation rather than recursion. We can add this to the TODO list.

**Comment by adam2392** [comment]
Thanks @aryan26roy !

**Pull request merged**